### PR TITLE
[WorkInProgress] Memory Profiling

### PR DESCRIPTION
**Pull request by dendibakh**
In this section, I want to cover:
* basics: memory consumption over time, RSS, heap allocations
* advanced: memory footprint over time (hot memory regions per unit of time)
* [maybe] very advanced: analyze access patterns, temporal and spatial locality, reuse distance.

### CODE REVIEW

File: chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+### Case Study: Memory Usage of Stockfish
+
+Now, let's take a look at how to profile the memory usage of a real-world application. We will use [heaptrack](https://github.com/KDE/heaptrack)[^2], an open-sourced heap memory profiler for Linux developed by KDE. Ubuntu users can install it very easily with `apt install heaptrack heaptrack-gui`. Heaptrack can find places in the code where the largest and most frequent allocations happen among many other things. On Windows, you can use [Mtuner](https://github.com/milostosic/MTuner)[^3] which has similar capabilities as Heaptrack.
+
+As an example, we took Stockfish's built-in benchmark, which we have already analyzed in Chapter 4. As before, we compiled it using the Clang 15 compiler with `-O3 -mavx2` options. We collected the Heaptrack memory profile of a single-threaded Stockfish built-in benchmark on an Intel Alderlake i7-1260P processor using the following command:
```
Comment (olest) [inline chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md:54]: URL or reference?

File: chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+We will show how to obtain such charts in the following two case studies, but for now, you may wonder how this data can be used. Well, first, if we sum up unique bytes (dotted lines) accessed during every interval, we will get the total memory footprint of a program. Also, by looking at the chart, you can observe phases and correlate them with the code that is running. Ask yourself: "Does it look according to your expectations, or the workload is doing something sneaky?" You may encounter unexpected spikes in memory footprint. Memory profiling techniques that we will discuss in this section do not necessarily point you to the problematic places similar to regular hotspot profiling but they certainly help you better understand the behavior of a workload. On many occasions, memory profiling helped identify a problem or served as an additional data point to support the conclusions that were made during regular profiling.
```
[the remark this answers is not in the document]
Comment (dendibakh) [reply chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md:54]: Added a TODO in the text.

File: chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md (hunk 1 above)
Comment (dendibakh) [reply chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md]: Fixed.

File: chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+By default, SDE measures footprint in cache lines (64 KB), but it can also measure it in memory pages (4KB on x86). We combined the output and put it side by side. Also, a few non-relevant columns were removed from the output. The first column `PERIOD` marks the start of a new interval of 28K instructions. The difference between each period is 28K instructions. The column `LOAD` tells how many cache lines were accessed by load instructions. Recall from the previous discussion, the same cache line accessed twice counts only once. Similarly, the column `STORE` tells how many cache lines were stored. The column `CODE` counts accessed cache lines that contain instructions that were executed during that period. Finally, `NEW` counts cache lines touched during a period, that were not seen before by the program.
```
Comment (ilitov) [inline chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md:128]: I think you meant `cache lines (64B on x86)` and not `64KB` ? :)

Comment (dendibakh) [reply chapters/8-Optimizing-Memory-Accesses/8-5 Memory Profiling.md]: fixed. https://github.com/dendibakh/perf-book/pull/27/commits/0bf2cdcd7252445524faba9082ea155a08fd1dee

**Pull request merged**